include/stomp_moveit/trajectory_visualization.hpp

```diff
@@ -5,55 +5,171 @@
+  }
+
+  //RCLCPP_ERROR_STREAM(LOGGER, "get_subscription_count: " << marker_publisher->get_subscription_count()
+  //                                                       << " # of markers: " << markers_array.markers.size());
+  marker_publisher->publish(markers_array);
+  return true;
+}
+
 PostIterationFn
-get_iteration_path_publisher(moveit_visual_tools::MoveItVisualTools& visual_tools,
-                             const moveit::core::JointModelGroup* group,
-                             const rviz_visual_tools::Colors color = rviz_visual_tools::TRANSLUCENT_LIGHT)
+get_iteration_path_publisher(rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_publisher,
+                             planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor,
+                             const moveit::core::JointModelGroup* group)
 {
   assert(group != nullptr);
 
   std::shared_ptr<const moveit::core::RobotState> reference_state;
   {
-    planning_scene_monitor::LockedPlanningSceneRO scene(visual_tools.getPlanningSceneMonitor());
+    planning_scene_monitor::LockedPlanningSceneRO scene(planning_scene_monitor);
     reference_state = std::make_shared<moveit::core::RobotState>(scene->getCurrentState());
   }
 
-  PostIterationFn path_publisher = [=, &visual_tools](int /*iteration_number*/, double /*cost*/,
-                                                      const Eigen::MatrixXd& values) {
-    static thread_local robot_trajectory::RobotTrajectory trajectory(visual_tools.getRobotModel(), group);
+  PostIterationFn path_publisher = [=](int /*iteration_number*/, double /*cost*/, const Eigen::MatrixXd& values) {
+    static thread_local robot_trajectory::RobotTrajectory trajectory(planning_scene_monitor->getRobotModel(), group);
     fill_robot_trajectory(values, *reference_state, trajectory);
-    visual_tools.publishTrajectoryLine(trajectory, group, color);
-    visual_tools.trigger();
+
+    std::vector<const moveit::core::LinkModel*> tips;
+    if (!group->getEndEffectorTips(tips))
+    {
+      RCLCPP_ERROR_STREAM(LOGGER, "Unable to get end effector tips from jmg");
+    }
+
+    // For each end effector
+    for (const moveit::core::LinkModel* ee_parent_link : tips)
+    {
+      publishTrajectoryPoints(trajectory, ee_parent_link, marker_publisher, TRANSLUCENT_LIGHT);
+    }
   };
 
   return path_publisher;
 }
 
-DoneFn get_success_trajectory_publisher(moveit_visual_tools::MoveItVisualTools& visual_tools,
-                                        const moveit::core::JointModelGroup* group,
-                                        const rviz_visual_tools::Colors color = rviz_visual_tools::GREEN)
+DoneFn
+get_success_trajectory_publisher(rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_publisher,
+                                 planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor,
+                                 const moveit::core::JointModelGroup* group)
```

Comment:
This function seems to be highly redundant to `get_iteration_path_publisher`. Are there any differences at all?

Reply:
It's really only a separate callback in STOMP https://github.com/ros-industrial/stomp/blob/main/include/stomp/task.h#L160